store/src/server/api.rs

```diff
@@ -1,27 +1,104 @@
+        request: tonic::Request<Streaming<store::CheckNullifiersRequest>>,
     ) -> Result<Response<Self::CheckNullifiersStream>, Status> {
-        todo!()
+        let mut messages = request.into_inner();
+
+        let (tx, rx) = mpsc::channel(128);
+        let tree = self.nullifier_tree.clone();
+
+        // spawn a separate task so that user disconnects can be detected and logged
+        tokio::spawn(async move {
+            // Consume messages sent by the user, when there are no more messages None is returned.
+            while let Some(msg) = messages.next().await {
+                // Stop processing on first error.
+                let nullifier = match msg {
+                    Err(status) => {
+                        error!(
+                            code = status.code().description(),
+                            msg = status.message(),
+                            "User sent an invalid message"
+                        );
+                        return Err(status);
+                    },
+                    Ok(store::CheckNullifiersRequest { nullifier: None }) => {
+                        error!("Received empty nullifier");
+                        return Err(Status::invalid_argument(
+                            "CheckNullifiersRequest received the empty nullifier",
+                        ));
+                    },
+                    Ok(store::CheckNullifiersRequest {
+                        nullifier: Some(nullifier),
+                    }) => nullifier,
+                };
+
+                if nullifier.d0 >= <Felt as StarkField>::MODULUS {
+                    return Err(Status::invalid_argument(
+                        "Digest field d0 is not in the modulos range",
+                    ));
+                }
+                if nullifier.d1 >= <Felt as StarkField>::MODULUS {
+                    return Err(Status::invalid_argument(
+                        "Digest field d1 is not in the modulos range",
+                    ));
+                }
+                if nullifier.d2 >= <Felt as StarkField>::MODULUS {
+                    return Err(Status::invalid_argument(
+                        "Digest field d2 is not in the modulos range",
+                    ));
+                }
+                if nullifier.d3 >= <Felt as StarkField>::MODULUS {
+                    return Err(Status::invalid_argument(
+                        "Digest field d3 is not in the modulos range",
+                    ));
+                }
+
+                let digest = RpoDigest::new([
+                    Felt::new(nullifier.d0),
+                    Felt::new(nullifier.d1),
+                    Felt::new(nullifier.d2),
+                    Felt::new(nullifier.d3),
+                ]);
+                let leaf = tree.read().await.get_value(digest);
+
+                tx.send(Ok(CheckNullifiersResponse {
+                    result: leaf != TieredSmt::EMPTY_VALUE,
+                }))
+                .await
```

Comment:
Here, we should also be sending a Merkle path from the tree root to this nullifier.

Reply:
implemented